// ra-ui-materialui/lib/form/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFormInitialValues = exports.Toolbar = exports.TabbedFormView = exports.TabbedFormTabs = exports.SimpleFormView = exports.SimpleForm = exports.FormInput = void 0;
var FormInput_1 = __importDefault(require("./FormInput"));
exports.FormInput = FormInput_1.default;
var SimpleForm_1 = __importDefault(require("./SimpleForm"));
exports.SimpleForm = SimpleForm_1.default;
var TabbedFormTabs_1 = __importDefault(require("./TabbedFormTabs"));
exports.TabbedFormTabs = TabbedFormTabs_1.default;
var Toolbar_1 = __importDefault(require("./Toolbar"));
exports.Toolbar = Toolbar_1.default;
var getFormInitialValues_1 = __importDefault(require("./getFormInitialValues"));
exports.getFormInitialValues = getFormInitialValues_1.default;
var SimpleFormView_1 = require("./SimpleFormView");
Object.defineProperty(exports, "SimpleFormView", { enumerable: true, get: function () { return SimpleFormView_1.SimpleFormView; } });
var TabbedFormView_1 = require("./TabbedFormView");
Object.defineProperty(exports, "TabbedFormView", { enumerable: true, get: function () { return TabbedFormView_1.TabbedFormView; } });
__exportStar(require("./SimpleFormIterator"), exports);
__exportStar(require("./TabbedForm"), exports);
__exportStar(require("./FormTab"), exports);
__exportStar(require("./FormTabHeader"), exports);
