// ra-core/esm/form/useInput.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { useCallback, useEffect } from 'react';
import { useField as useFinalFormField, useForm, } from 'react-final-form';
import { composeValidators } from './validate';
import isRequired from './isRequired';
import { useFormGroupContext } from './useFormGroupContext';
import { useFormContext } from './useFormContext';
import { useRecordContext } from '../controller';
var useInput = function (_a) {
    var defaultValue = _a.defaultValue, initialValue = _a.initialValue, id = _a.id, name = _a.name, source = _a.source, validate = _a.validate, customOnBlur = _a.onBlur, customOnChange = _a.onChange, customOnFocus = _a.onFocus, isRequiredOption = _a.isRequired, options = __rest(_a, ["defaultValue", "initialValue", "id", "name", "source", "validate", "onBlur", "onChange", "onFocus", "isRequired"]);
    var finalName = name || source;
    var formGroupName = useFormGroupContext();
    var formContext = useFormContext();
    var record = useRecordContext();
    useEffect(function () {
        if (!formContext || !formGroupName) {
            return;
        }
        formContext.registerField(source, formGroupName);
        return function () {
            formContext.unregisterField(source, formGroupName);
        };
    }, [formContext, formGroupName, source]);
    var sanitizedValidate = Array.isArray(validate)
        ? composeValidators(validate)
        : validate;
    var _b = useFinalFormField(finalName, __assign({ initialValue: initialValue,
        defaultValue: defaultValue, validate: sanitizedValidate }, options)), input = _b.input, meta = _b.meta;
    // Extract the event handlers so that we can provide ours
    // allowing users to provide theirs without breaking the form
    var onBlur = input.onBlur, onChange = input.onChange, onFocus = input.onFocus, inputProps = __rest(input, ["onBlur", "onChange", "onFocus"]);
    var handleBlur = useCallback(function (event) {
        onBlur(event);
        if (typeof customOnBlur === 'function') {
            customOnBlur(event);
        }
    }, [onBlur, customOnBlur]);
    var handleChange = useCallback(function (event) {
        onChange(event);
        if (typeof customOnChange === 'function') {
            customOnChange(event);
        }
    }, [onChange, customOnChange]);
    var handleFocus = useCallback(function (event) {
        onFocus(event);
        if (typeof customOnFocus === 'function') {
            customOnFocus(event);
        }
    }, [onFocus, customOnFocus]);
    var form = useForm();
    var recordId = record === null || record === void 0 ? void 0 : record.id;
    // Every time the record changes and doesn't include a value for this field,
    // reset the field value to the initialValue (or defaultValue)
    useEffect(function () {
        if (typeof input.checked !== 'undefined' || // checkbox that has a value from record
            (input.value != null && input.value !== '') // any other input that has a value from record
        ) {
            // no need to apply a default value
            return;
        }
        // Apply the default value if provided
        // We use a change here which will make the form dirty but this is expected
        // and identical to what FinalForm does (https://final-form.org/docs/final-form/types/FieldConfig#defaultvalue)
        if (defaultValue != null) {
            form.change(source, defaultValue);
        }
        // apply initial value if provided
        if (initialValue != null) {
            form.batch(function () {
                form.change(source, initialValue);
                form.resetFieldState(source);
            });
        }
    }, [
        recordId,
        input.value,
        input.checked,
        defaultValue,
        initialValue,
        source,
        form,
    ]);
    // If there is an input prop, this input has already been enhanced by final-form
    // This is required in for inputs used inside other inputs (such as the SelectInput inside a ReferenceInput)
    if (options.input) {
        return {
            id: id || source,
            input: options.input,
            meta: options.meta,
            isRequired: isRequiredOption || isRequired(validate),
        };
    }
    return {
        id: id || source,
        input: __assign(__assign({}, inputProps), { onBlur: handleBlur, onChange: handleChange, onFocus: handleFocus }),
        meta: meta,
        isRequired: isRequiredOption || isRequired(validate),
    };
};
export default useInput;
