// ra-core/lib/inference/assertions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.valuesAreObject = exports.isObject = exports.valuesAreDateString = exports.isDateString = exports.valuesAreDate = exports.isDate = exports.valuesAreArray = exports.isArray = exports.valuesAreEmail = exports.isEmail = exports.valuesAreImageUrl = exports.isImageUrl = exports.valuesAreUrl = exports.isUrl = exports.valuesAreHtml = exports.isHtml = exports.valuesAreString = exports.isString = exports.valuesAreBooleanString = exports.isBooleanString = exports.valuesAreBoolean = exports.isBoolean = exports.valuesAreInteger = exports.isInteger = exports.valuesAreNumeric = exports.isNumeric = void 0;
var is_valid_1 = __importDefault(require("date-fns/is_valid"));
var parse_1 = __importDefault(require("date-fns/parse"));
exports.isNumeric = function (value) {
    return !isNaN(parseFloat(value)) && isFinite(value);
};
exports.valuesAreNumeric = function (values) { return values.every(exports.isNumeric); };
exports.isInteger = function (value) {
    return Number.isInteger(value) || !isNaN(parseInt(value));
};
exports.valuesAreInteger = function (values) { return values.every(exports.isInteger); };
exports.isBoolean = function (value) { return typeof value === 'boolean'; };
exports.valuesAreBoolean = function (values) { return values.every(exports.isBoolean); };
exports.isBooleanString = function (value) {
    return ['true', 'false'].includes(value.toString().toLowerCase());
};
exports.valuesAreBooleanString = function (values) {
    return values.every(exports.isBooleanString);
};
exports.isString = function (value) { return typeof value === 'string'; };
exports.valuesAreString = function (values) { return values.every(exports.isString); };
var HtmlRegexp = /<([A-Z][A-Z0-9]*)\b[^>]*>(.*?)<\/\1>/i;
exports.isHtml = function (value) { return !value || HtmlRegexp.test(value); };
exports.valuesAreHtml = function (values) { return values.every(exports.isHtml); };
var UrlRegexp = /http(s*):\/\/.*/i;
exports.isUrl = function (value) { return !value || UrlRegexp.test(value); };
exports.valuesAreUrl = function (values) { return values.every(exports.isUrl); };
var ImageUrlRegexp = /http(s*):\/\/.*\.(jpeg|jpg|jfif|pjpeg|pjp|png|svg|gif|webp|apng|bmp|ico|cur|tif|tiff)/i;
exports.isImageUrl = function (value) { return !value || ImageUrlRegexp.test(value); };
exports.valuesAreImageUrl = function (values) { return values.every(exports.isImageUrl); };
// This is a very simple regex to find emails
// It it NOT meant to validate emails as the spec is way more complicated but is
// enough for our inference needs
var EmailRegexp = /@{1}/;
exports.isEmail = function (value) { return !value || EmailRegexp.test(value); };
exports.valuesAreEmail = function (values) { return values.every(exports.isEmail); };
exports.isArray = function (value) { return Array.isArray(value); };
exports.valuesAreArray = function (values) { return values.every(exports.isArray); };
exports.isDate = function (value) { return !value || value instanceof Date; };
exports.valuesAreDate = function (values) { return values.every(exports.isDate); };
exports.isDateString = function (value) {
    return !value || (typeof value === 'string' && is_valid_1.default(parse_1.default(value)));
};
exports.valuesAreDateString = function (values) {
    return values.every(exports.isDateString);
};
exports.isObject = function (value) {
    return Object.prototype.toString.call(value) === '[object Object]';
};
exports.valuesAreObject = function (values) { return values.every(exports.isObject); };
